appserver/web/web-core/src/main/java/org/apache/catalina/core/StandardContextValve.java

```diff
@@ -286,13 +287,19 @@ protected String normalize(String path) {
 
         String rv = path;
         // starts with a double-slash
-        if(rv.indexOf("//") == 0) {
+        if (rv.indexOf("//") == 0) {
             rv = rv.replace("//", "/");
         }
         // starts with dot-slash
-        if(rv.indexOf("./") == 0) {
+        if (rv.indexOf("./") == 0) {
             rv = rv.replaceFirst("./", "/");
         }
+        // has WEB-INF or META-INF
+        if (rv.toUpperCase().indexOf("WEB-INF") > 0) {
+            rv = "/" + rv.substring(rv.toUpperCase().indexOf("WEB-INF"));
+        } else if (rv.toUpperCase().indexOf("META-INF") > 0) {
+            rv = "/" + rv.substring(rv.toUpperCase().indexOf("META-INF"));
+        }
```

Comment:
How hot is this path, what percentage of requests would hit it? If almost every one, then two uppercase conversions is a little bit wasteful and case-insentivive regexp would be more efficient, though less readable.

Reply:
Performance improved and now checking if it contains "/WEB-INF/" or ends with "/WEB-INF".
The same check for "META-INF